model/src/main/kotlin/utils/FindingsMatcher.kt

```diff
@@ -274,5 +280,67 @@ fun associateLicensesWithExceptions(
         exception.copy(license = "${SpdxConstants.NOASSERTION} ${SpdxExpression.WITH} ${exception.license}".toSpdx())
     }
 
-    return fixedLicenses
+    return fixedLicenses.mapTo(mutableSetOf()) { it.copy(license = associateLicensesWithExceptions(it.license)) }
+}
+
+/**
+ * Process [license] for stand-alone license exceptions as part of compound expressions and associate them with
+ * applicable licenses. Orphan license exceptions will get associated by [SpdxConstants.NOASSERTION]. Return a new
+ * expression that does not contain stand-alone license exceptions anymore.
+ */
+internal fun associateLicensesWithExceptions(license: SpdxExpression): SpdxExpression {
+    // If this is not a compound expression, there can be no stand-alone license exceptions with belonging licenses.
+    if (license !is SpdxCompoundExpression) return license
+
+    // Exclusively operate on AND-only expressions without further nested expressions.
+    val hasOnlyAndOperator = license.operator == SpdxOperator.AND && "(" !in license.toString()
+    if (!hasOnlyAndOperator) {
+        return SpdxCompoundExpression(
+            associateLicensesWithExceptions(license.left),
+            license.operator,
+            associateLicensesWithExceptions(license.right)
+        )
+    }
+
+    val associatedLicenses = mutableSetOf<SpdxSingleLicenseExpression>()
+    val remainingLicenses = mutableSetOf<SpdxSimpleExpression>()
+    val remainingExceptions = mutableSetOf<SpdxSingleLicenseExpression>()
+
+    // Divide the AND-operands into exceptions, simple expressions, and licenses than cannot be used with an exception.
+    license.decompose().forEach {
+        when {
+            SpdxLicenseException.forId(it.toString()) != null -> remainingExceptions += it
+            it is SpdxSimpleExpression -> remainingLicenses += it
+            else -> associatedLicenses += it
+        }
+    }
+
+    val i = remainingExceptions.iterator()
+
+    while (i.hasNext()) {
+        val exception = i.next()
+        val exceptionString = exception.toString()
+
+        // Determine all licenses the exception is applicable to.
+        val applicableLicenses = SpdxLicenseException.mapping[exceptionString].orEmpty().mapTo(mutableSetOf()) {
+            SpdxLicenseIdExpression(it.id)
+        }
+
+        // Associate all remaining licenses that are applicable with the exception.
+        remainingLicenses.intersect(applicableLicenses).forEach {
+            associatedLicenses += SpdxLicenseWithExceptionExpression(it, exceptionString)
+            remainingLicenses -= it
+            i.remove()
```

Comment:
Don't know much about SPDX processing, but is it correct that the `i.remove()` call is done inside the `forEach` loop? Is it guaranteed that it contains enough elements?

Reply:
Good finding! I'll add a test for this case.